presto-druid/src/main/java/com/facebook/presto/druid/DruidClient.java

```diff
@@ -83,6 +102,46 @@ public DruidClient(DruidConfig config, @ForDruidClient HttpClient httpClient)
         this.druidCoordinator = URI.create(config.getDruidCoordinatorUrl());
         this.druidBroker = URI.create(config.getDruidBrokerUrl());
         this.druidSchema = config.getDruidSchema();
+        this.caseInsensitiveNameMatching = config.isCaseInsensitiveNameMatching();
+
+        Duration caseInsensitiveNameMatchingCacheTtl = requireNonNull(config.getCaseInsensitiveNameMatchingCacheTtl(), "caseInsensitiveNameMatchingCacheTtl is null");
+        CacheBuilder<Object, Object> remoteTableNamesCacheBuilder = CacheBuilder.newBuilder()
+                .expireAfterWrite(caseInsensitiveNameMatchingCacheTtl.toMillis(), MILLISECONDS);
+        this.remoteTables = remoteTableNamesCacheBuilder.build();
+    }
+
+    Optional<RemoteTableObject> toRemoteTable(SchemaTableName schemaTableName)
+    {
+        requireNonNull(schemaTableName, "schemaTableName is null");
+        verify(CharMatcher.forPredicate(Character::isUpperCase).matchesNoneOf(schemaTableName.getTableName()), "Expected table name from internal metadata to be lowercase: %s", schemaTableName);
+        if (!caseInsensitiveNameMatching) {
+            return Optional.of(RemoteTableObject.of(schemaTableName.getTableName()));
+        }
+
+        @Nullable Optional<RemoteTableObject> remoteTable = remoteTables.getIfPresent(schemaTableName);
+        if (remoteTable != null) {
+            return remoteTable;
+        }
+
+        // Cache miss, reload the cache
+        Map<SchemaTableName, Optional<RemoteTableObject>> mapping = new HashMap<>();
+        for (String table : getTables()) {
+            SchemaTableName cacheKey = new SchemaTableName(getSchema(), table);
+            mapping.merge(
+                    cacheKey,
+                    Optional.of(RemoteTableObject.of(table)),
+                    (currentValue, collision) -> currentValue.map(current -> current.registerCollision(collision.get().getOnlyRemoteTableName())));
+            remoteTables.put(cacheKey, mapping.get(cacheKey));
+        }
+
+        // explicitly cache if the requested table doesn't exist
+        if (!mapping.containsKey(schemaTableName)) {
+            mapping.put(schemaTableName, Optional.empty());
```

Comment:
I see, you're using Optional.empty to indicate the non-existing table name.  Now I understand why you're using an Optional as the V.  

But we don't have to update mapping here, am I right? if the mapping does not contains the key, we can just do remoteTables.put(schemaTableName,, Optional.empty() and then return Optional.empty(). the logic should be identical I think

Reply:
@beinan I agree, I have made the changes as suggested by you.